src/tinygp/kernels/quasisep.py

```diff
@@ -635,134 +638,362 @@ def transition_matrix(self, X1: JAXArray, X2: JAXArray) -> JAXArray:
+        beta_quads = jnp.atleast_1d(beta_quads)
+        beta_mult = jnp.atleast_1d(beta_mult)
+
+        alpha = carma_quads2poly(jnp.append(alpha_quads, jnp.array([1.0])))[:-1]
+        beta = carma_quads2poly(jnp.append(beta_quads, beta_mult))
+
+        return cls.init(alpha, beta)
+
     def design_matrix(self) -> JAXArray:
-        p = self.alpha.shape[0]
-        return jnp.concatenate((jnp.eye(p - 1, p, k=1), -self.alpha[None]))
+        # for real exponential components
+        dm_real = jnp.diag(self.arroots.real * self._real_mask)
+
+        # for complex exponential components
+        dm_complex_diag = jnp.diag(self.arroots.real * self._complex_mask)
+
+        # upper triangle entries
+        dm_complex_u = jnp.diag((self.arroots.imag * self._complex_select)[:-1], k=1)
+
+        return dm_real + dm_complex_diag + -dm_complex_u.T + dm_complex_u
 
     def stationary_covariance(self) -> JAXArray:
-        return self.stn
+        p = self.acf.shape[0]
+
+        # for real exponential components
+        diag = jnp.diag(jnp.where(self.acf.real > 0, jnp.ones(p), -jnp.ones(p)))
+
+        # for complex exponential components
+        denom = jnp.where(self._real_mask, 1.0, self.arroots.imag)
+        diag_complex = jnp.diag(
+            2
+            * jnp.square(
+                self.arroots.real
+                / denom
+                * jnp.roll(self._complex_select, 1)
+                * self._complex_mask
+            )
+        )
+        c_over_d = self.arroots.real / denom
+
+        # upper triangular entries
+        sc_complex_u = jnp.diag((-c_over_d * self._complex_select)[:-1], k=1)
+
+        return diag + diag_complex + sc_complex_u + sc_complex_u.T
 
     def observation_model(self, X: JAXArray) -> JAXArray:
-        return jnp.append(
-            self.beta, jnp.zeros(self.alpha.shape[0] - self.beta.shape[0])
-        )
+        return self.obsmodel
 
     def transition_matrix(self, X1: JAXArray, X2: JAXArray) -> JAXArray:
         dt = X2 - X1
-        return (self.proj_inv @ (jnp.exp(self.roots * dt)[:, None] * self.proj)).real
+        c = -self.arroots.real
+        d = -self.arroots.imag
+        decay = jnp.exp(-c * dt)
+        sin = jnp.sin(d * dt)
+
+        tm_real = jnp.diag(decay * self._real_mask)
+        tm_complex_diag = jnp.diag(decay * jnp.cos(d * dt) * self._complex_mask)
+        tm_complex_u = jnp.diag(
+            (decay * sin * self._complex_select)[:-1],
+            k=1,
+        )
 
+        return tm_real + tm_complex_diag + -tm_complex_u.T + tm_complex_u
 
-def _prod_helper(a1: JAXArray, a2: JAXArray) -> JAXArray:
-    i, j = np.meshgrid(np.arange(a1.shape[0]), np.arange(a2.shape[0]))
-    i = i.flatten()
-    j = j.flatten()
-    if a1.ndim == 1:
-        return a1[i] * a2[j]
-    elif a1.ndim == 2:
-        return a1[i[:, None], i[None, :]] * a2[j[:, None], j[None, :]]
-    else:
-        raise NotImplementedError
+
+@jax.jit
+def carma_roots(poly_coeffs: JAXArray) -> JAXArray:
```

Comment:
I moved all the `staticmethod`s out to the top level - totally subjective, but I think I'd prefer to have them here!

Reply:
Totally fine with me!